File: sliding_window_maximum.cpp
// Question Link - https://www.interviewbit.com/problems/sliding-window-maximum/
// Solution

vector<int> Solution::slidingMaximum(const vector<int> &A, int B) {
    deque<int> deq;
    for(int i=0; i<B; i++)
    {
        while((!deq.empty())&&(A[deq.back()]<=A[i])) deq.pop_back();
        deq.push_back(i);
    }
    vector<int> ans;
    for(int i=B; i<A.size(); i++)
    {
        ans.push_back(A[deq.front()]);
        while((!deq.empty())&&(deq.front()<=i-B)) deq.pop_front();
        while((!deq.empty())&&(A[deq.back()]<=A[i])) deq.pop_back();
        deq.push_back(i);
    }
    ans.push_back(A[deq.front()]);
    return ans;
}

File: striver_sheet/stack_and_queue_part_II/sliding_window_maximum.py
# Leetcode solution
# Link : https://leetcode.com/problems/sliding-window-maximum/

from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        slider = []
        dq = deque()
        for i in range(k):
            while len(dq) > 0 and nums[dq[-1]] < nums[i]:
                dq.pop()
            dq.append(i)
        slider.append(nums[dq[0]])
        for i in range(k, len(nums)):
            while len(dq) > 0 and i - dq[0] > k-1:
                dq.popleft()
            while len(dq) > 0 and nums[dq[-1]] < nums[i]:
                dq.pop()
            dq.append(i)
            slider.append(nums[dq[0]])
        return slider
